### ISSUE TITLE
Refresh a token (Multiple Oauth2Clients)

### ISSUE THREAD

**Issue by supertiger1234** [googleapis/google-api-nodejs-client#1611]
So, i was reading the docs and i saw the only way to refresh a token is this
```
oauth2client.on('tokens', (tokens) => {
  if (tokens.refresh_token) {
    // store the refresh_token in my database!
    console.log(tokens.refresh_token);
  }
  console.log(tokens.access_token);
});
```
Isn't it inefficient to add an event to every single client? isn't there a better way to manually check the clients access token and refresh it using the refresh token?

**Comment by AVaksman** [comment]
@supertiger1234 
The refresh_token is only provided on the first authorization from the user. Subsequent authorizations, such as the kind you make while testing an OAuth2 integration, will not return the refresh_token again. :)

`access_token` is good for one hour. If your `access_token` is close to be expired or expired already the next time you execute, than your token will be refreshed automatically with the use of `refresh_token`.
By adding an listener on `tokens` event merely helps you record the new `access_token` value locally when it is refreshed automatically.

Does that help you?

**Comment by supertiger1234** [comment]
Ohhh, so it happens automatically?
Also, what if I wanna get the refresh token from a database? Do I just do setCredentials({access_token: db.accessToken})?

Thanks a lot for the quick reply

**Comment by fabiofcferreira** [comment]
Hey! I randomly came across this issue and I actually wanted to ask something very similar to @supertiger1234 asked.

What if I want to save the new token to the database? Using the listener becomes very hard in that matter.
What is your suggestion?

**Comment by mrdulin** [comment]
Same issue. I want to ask what's the difference about these two ways:

Way 1: `oauth2client.setCredentials({access_token: db.accessToken})`

Way 2: `oauth2client.setCredentials({refresh_token: db.refreshToken})`

According to the docs, I think the way 2 is enough. `oauth2client` will always refresh access token **automatically**.

**Comment by Shivamdot** [comment]
oauth2client.on('tokens', (tokens) => {
  //store access_token in your database
  console.log(tokens.access_token); // for testing
});

this function "oauth2client.on('tokens')" will only run(automatically), when your previous access_token gets expired.
So inside this function, you can get the latest access_token and store it onto your database. 
Storing new access_token is important in order to prevent unnecessary requests for a new access token.

**Comment by bcoe** [comment]
@mrdulin @fabiofcferreira for the use-case where credentials are being stored for a long period of time, and you're relying on the `refresh_token`.

I would recommend storing both the `access_token` and the `refresh_token`, and calling `oauth2client.setCredentials` with both, e.g.,

```js
oauth2client.setCredentials({refresh_token: db.refreshToken, access_token: db.accessToken})
```

To make sure you have stored the most recent version of the `access_token`, you will need to use `.on("tokens")`. There's some useful examples of this [pattern in action here](oauth2client.setCredentials).

I hope this helps clarify 👍 

---

I'm going to go ahead and close this issue, since I think @supertiger1234 is unblocked, _but_ if other folks have follow up questions, please don't hesitate to open an additional issue.

**Comment by richardpeng** [comment]
@bcoe could you update the link to examples of the pattern for storing the `access_token` and `refresh_token`. I can't for the life of me figure out how to store my tokens because I don't have access to the user context when using `auth.on('tokens', (tokens) => {})`. If not something from the user context, what would you use as the key for storing tokens in a database?

**Comment by mrdulin** [comment]
@richardpeng I agree. 

The `tokens` event can't access application context. Where should I put this `.on("tokens")` in application code?

For example, if use `express`, the context info is attached in `req` object, like `req.user`

I think below way is wrong:

```js
app.post('/post', (req, res) => {
  auth.on('tokens', async (tokens) => {
    const requestingUser = req.user
    const user = await db.findByEmail(requestingUser.email)
    await db.updateGoogleAccountTokensByUser(user)
  })
})
```

**Comment by bfredo123** [comment]
Like @mrdulin and @richardpeng I would also like to have a solution to manage the tokens refresh for several users (stored in a DB).
 
Also, what is not clear to me, is what happens if the server is shut down and restarted. I guess the automatic refresh works because google-api-nodejs-client keeps in RAM a list of the tokens to be refreshed. But if this is right, then the RAM is lost whenever the server is restarted. 
Then, how to prevent all users to re-authenticate again upon every server upgrade?

**Comment by bytrangle** [comment]
Can anyone tell me when am I supposed to call this `.on('tokens')` event? Is it the first time the user authenticates? Because as far as I understand, if this event is called in subsequent API call, it won't return the `refresh_token`. And I've discovered that in subsequent API call, I only need to attach the refresh token:
```
oAuth2Client.setCredentials({
  refresh_token: REFRESH_TOKEN
})
```
to get a successful response. The Google API client automatically gets the new access token for me.

**Comment by iamrakhmatov** [comment]
@bfredo123 have u guys found a solution for your problems. I am currently having the same issue with storing access token using token event since cant get any user data inside of event handler

**Comment by ryordanov** [comment]
Hi all, I'm struggling with this too and found another option:
(1)
const oauth2 = google.oauth2('v2');
...........
oauth2Client.on('tokens', async (tokens) => {
          const tokenInfo = await oauth2.tokeninfo({ id_token: tokens.id_token });
          if (tokenInfo.data && tokenInfo.data.email) {
              db.refreshToken(tokens, tokenInfo.data.email);
          }
      });
This way you may find user in you db by email (but I guess it makes one more request to https://oauth2.googleapis.com/tokeninfo?id_token=XXX)

Not sure which approach is better - additional request(1) or callback overwrite on every request(2), but at least we have choice :)
(2)
app.post('/post', (req, res) => {
  const userId = req.user.id;
  oauth2Client.on('tokens', async (tokens) => {tokens, userId, ...})
}

**Comment by Shivamdot** [comment]
> @richardpeng @mrdulin have u guys found a solution for your problems. I am currently having the same issue with storing access token using token event since cant get any user data inside of event handler

There is no function that will give you user info or user ID. It's just u need a way to identify your users uniquely, maybe via a login system, just like when u are identifying an access token for a particular user.

**Comment by SteveShengStar** [comment]
I have an inquiry also related to declaring oauth2Client as a variable globally in a JS file vs. locally inside a function. Replica of this Stack Overflow post I created: https://stackoverflow.com/questions/68145083/oauth2-globally-defined-client-concurrent-backend-function-execution

I have some confusion about integrating OAuth 2 flow (Google APIs) into my app. I have declared a Oauth2 client variable (globally) as shown below:

const {gapi} = require('googleapis');
var auth2;

/**
 * Initializes the Sign-In client.
 */
var initClient = function() {
    gapi.load('auth2', function(){
        /**
         * Retrieve the singleton for the GoogleAuth library and set up the
         * client.
         */
        auth2 = gapi.auth2.init({
            client_id: 'CLIENT_ID.apps.googleusercontent.com'
        });

        // Attach the click handler to the sign-in button
        auth2.attachClickHandler('signin-button', {}, onSuccess, onFailure);
    });
};
If I were to use Google Calendar API, I would do this:

const doSomethingWithGCalendar = (token) => {
    auth2.setCredentials({
       access_token: token
    });
        
    const calendar = google.calendar({version: 'v3', auth: auth2});
// Execute some actions here
}
I am confused about how node.js handles this. After all, the auth2 client variable is globally declared (I picture in my head as one, single instance). However, there will be multiple users using my app, so the line .setCredentials and subsequent Google Calendar API calls will run, probably in multiple threads and concurrently, I'm assuming.

They should be running concurrently in the node.js server, but they're using the same globally declared variable auth2. How would that work ?

Should I, instead, declare a local variable inside doSomethingWithGCalendar() like this:

const doSomethingWithGCalendar = (token) => {
    const auth2 = new OAuth2Client(
        {client_id: 'CLIENT_ID.apps.googleusercontent.com'}
    );
    // Set Credentials/tokens and then run Google Calendar API operations ...
